subscriber.go

```diff
@@ -811,11 +885,81 @@ func (h *handler) handle(ctx context.Context, nextCid cid.Cid, sel ipld.Node, wr
 		return nil, nil
 	}
 
-	err := syncer.Sync(ctx, nextCid, sel)
-	if err != nil {
-		return nil, err
+	var limitOK bool
+	var origLimit selector.RecursionLimit
+	// Only attempt to extract original recursion limit from selector if segmentation is enabled
+	// and there is a segment block hook.
+	//
+	// Not that we need the action in segment block hook to know which CID to sync in next segment.
+	// Therefore, it has to be set for segmented sync to function correctly.
+	if h.subscriber.segDepthLimit < 0 && h.subscriber.segBlockHook != nil {
+		origLimit, limitOK = getRecursionLimit(sel)
+	}
+
+	// Revert back to sync without segmentation if original limit was not extracted, due to:
+	// - segment depth limit being negative; meaning segmentation is explicitly disabled, or
+	// - no segment block hook is configured; meaning we don't have a way to determin next
+	//   CID during segmented sync, or
+	// - the selector does not explore recursively, and has no top level recursion limit.
+	if !limitOK {
```

Comment:
The idea here is that segmented sync will only happen if it is enabled via config, i.e. both hook and depth for it are specified _and_ the original selector given to sync the dag with explores recursively with some recursion limit.

This is again on the conservative side to avoid accidental syncing of CIDs that the original selector might not have selected.